// @fetcch/id/dist/types/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiError = exports.PaymentRequestNotFound = exports.ParamsNotCorrect = exports.ApiKeyNotValid = exports.IdNotFound = void 0;
class IdNotFound extends Error {
    constructor(message) {
        super(`${message} id associated with this was not found in arweave nor in database`);
        this.statusCode = 400;
    }
}
exports.IdNotFound = IdNotFound;
class ApiKeyNotValid extends Error {
    constructor(message) {
        super(`api_key is not valid`);
        this.statusCode = 401;
    }
}
exports.ApiKeyNotValid = ApiKeyNotValid;
class ParamsNotCorrect extends Error {
    constructor(message) {
        super(message);
        this.statusCode = 400;
    }
}
exports.ParamsNotCorrect = ParamsNotCorrect;
class PaymentRequestNotFound extends Error {
    constructor(message) {
        super(message);
        this.statusCode = 400;
    }
}
exports.PaymentRequestNotFound = PaymentRequestNotFound;
class ApiError extends Error {
    constructor(message, statusCode) {
        super(message);
        this.statusCode = 400;
        this.statusCode = statusCode;
    }
}
exports.ApiError = ApiError;
